Fix empty message when host legacy routing is true

vincentmli opened the pull request:
when missing tunnel and socketLB service check and  when legacy host routing
set to true, the msg is empty with log

"level=warning msg=" Falling back to iptables-based masquerading." subsys=daemon"

this is confusing to users and not sure why falling back to iptables-based masquerading

before the fix:

run cilium agent in kernel 4.19 with

1 BPF masq enabled
2 Host legacy routing enabled
3 Tunnel enabled
4 FullHostReachableServices false

Cilium agent logs:
 "level=warning msg=" Falling back to iptables-based masquerading." subsys=daemon"

the msg is empty because none of the switch cases matches, this could mis-lead users to
think BPF masquerade is not supported on kernel 4.19.

then when:

1 BPF masq enabled
2 Host legacy routing enabled
3 Tunnel disabled
4 Nodeport enabled

cilium agent keeps BPF masquerade enabled on kernel 4.19 without falling back to iptables-based masquerading

after the fix:

cilum agent log should log correct message + "Falling back to iptables-based masquerading."

Signed-off-by: Vincent Li <v.li@f5.com>

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #21237

```release-note
fix empty message when tunnel and socketLB service missing in switch case
```

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -930,6 +930,9 @@ func NewDaemon(ctx context.Context, cleaner *daemonCleanup, epMgr *endpointmanag
 			msg = fmt.Sprintf("BPF masquerade does not allow to specify devices via --%s (use --%s instead).",
 				option.EgressMasqueradeInterfaces, option.Devices)
+		case option.Config.EnableHostLegacyRouting:
+			msg = fmt.Sprintf("BPF masquerade requires full host reachable services enabled in tunnel mode but host legacy routing (--%s=\"true\")",
```
Host reachable services (a.k.a. socket-level load balancing) and legacy host routing are two different things.

I think it would also be better to refactor this code so we don't have to keep the switch cases in sync. with the above if condition.

vincentmli replied:
@brb it appears you are familiar with this part of code and the background, what do you think about the refactor?

vincentmli replied:
it looks we can merge the if checks block and the switch case checks so we don't need to check the options twice.

vincentmli replied:
I made an initial refactor, please check if that is ok, I could not test it locally as it appears `make dev-docker-image` stop working anymore probably due to some master branch changes

pchaigno replied:
> I could not test it locally as it appears `make dev-docker-image` stop working anymore probably due to some master branch changes

It works fine for me (on commit 1059b958c8e9f9c88a810218ccc928900e4c98df). Are you on latest master? If you are and it persists, could you open a thread on Slack?

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -921,9 +921,7 @@ func NewDaemon(ctx context.Context, cleaner *daemonCleanup,
-	if option.Config.EnableIPv4Masquerade && option.Config.EnableBPFMasquerade &&
-		(!option.Config.EnableNodePort || option.Config.EgressMasqueradeInterfaces != "" || !option.Config.EnableRemoteNodeIdentity ||
-			(option.Config.TunnelingEnabled() && !hasFullHostReachableServices())) {
+	if option.Config.EnableIPv4Masquerade && option.Config.EnableBPFMasquerade {
```
This refactoring will end up always disabling BPF masquerading and throwing a warning.

vincentmli replied:
ah, maybe add `if (msg)` then set `option.Config.EnableBPFMasquerade = false`

vincentmli replied:
I also don't need to add the switch `case option.Config.EnableHostLegacyRouting` after the refactor, since empty msg will not happen even I enable `option.Config.EnableHostLegacyRouting`

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -936,6 +934,13 @@ func NewDaemon(ctx context.Context, cleaner *daemonCleanup,
 			msg = fmt.Sprintf("BPF masquerade does not allow to specify devices via --%s (use --%s instead).",
 				option.EgressMasqueradeInterfaces, option.Devices)
+		case option.Config.TunnelingEnabled():
+			if !hasFullHostReachableServices() {
```
```suggestion
		case option.Config.TunnelingEnabled() && !hasFullHostReachableServices() {
```

vincentmli replied:
the  suggestion result in error `daemon/cmd/daemon.go|938 col 35| expected ':', found '{' (and 3 more errors)`, what about 
```
case option.Config.TunnelingEnabled() && !hasFullHostReachableServices():
                        msg = "BPF masquerade requires full host reachable services enabled in tunnel mode."
```

pchaigno replied:
Yep, that's what I meant to write :+1:

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -936,6 +934,13 @@ func NewDaemon(ctx context.Context, cleaner *daemonCleanup,
+			}
+		case option.Config.EnableHostLegacyRouting:
+			msg = fmt.Sprintf("BPF masquerade requires host legacy routing (--%s=\"false\")",
+				option.EnableHostLegacyRouting)
```
Why was this case added?

vincentmli replied:
this is when I set host legacy routing to true in my test, then non of the switch cases match, result in empty msg, then following `log.Warn(msg + " Falling back to iptables-based masquerading.")` logs empty msg + `Falling back to iptables-based masquerading`, user will be confused why falling back to iptables-based masquerading,  the `if !option.Config.EnableHostLegacyRouting {` at the end of the block check does not match, so the `log.Info` is missed too

pchaigno replied:
Are you sure this was related to legacy host routing? Could it not be because you had host reachable services disabled?

vincentmli replied:
after refactor and add `if msg`, does not make sense to add `option.Config.EnableHostLegacyRouting` switch case, the refactor seems not only improve the code, but also the logic, please check the new changes to see if it make sense

pchaigno reviewed (changes requested):
The code looks good to me!

The commit description needs to be updated to reflect latest changes. Could you please also separate refactoring from fix, i.e. in 2 commits?

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -939,14 +939,16 @@ func NewDaemon(ctx context.Context, cleaner *daemonCleanup,
-			option.Config.EnableHostLegacyRouting = true
-			log.Infof("BPF masquerade could not be enabled. Falling back to legacy host routing (--%s=\"true\").",
-				option.EnableHostLegacyRouting)
+		if len(msg) > 0 {
```
The check on `msg` belongs in the refactoring.

vincentmli replied:
ack

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -936,6 +934,8 @@ func NewDaemon(ctx context.Context, cleaner *daemonCleanup,
 			msg = fmt.Sprintf("BPF masquerade does not allow to specify devices via --%s (use --%s instead).",
 				option.EgressMasqueradeInterfaces, option.Devices)
+		case option.Config.TunnelingEnabled() && !hasFullHostReachableServices():
+			msg = "BPF masquerade requires full host reachable services enabled in tunnel mode."
```
This change belongs in the fix commit. You'll have to place the fix commit before the refactoring commit to avoid introducing a regression in between the two commits.

vincentmli replied:
yep, make sense

vincentmli committed:
Fix: Tunnel and sokcetLB service missing in switch case result in empty msg

when missing tunnel and socketLB service check and  when legacy host routing
set to true, the msg is empty with log

"level=warning msg=" Falling back to iptables-based masquerading." subsys=daemon"

this is confusing to users and not sure why falling back to iptables-based masquerading

before the fix:

run cilium agent in kernel 4.19 with

1 BPF masq enabled
2 Host legacy routing enabled
3 Tunnel enabled
4 FullHostReachableServices false

Cilium agent logs:
 "level=warning msg=" Falling back to iptables-based masquerading." subsys=daemon"

the msg is empty because none of the switch cases matches, this could mis-lead users to
think BPF masquerade is not supported on kernel 4.19.

then when:

1 BPF masq enabled
2 Host legacy routing enabled
3 Tunnel disabled
4 Nodeport enabled

cilium agent keeps BPF masquerade enabled on kernel 4.19 without falling back to iptables-based masquerading

after the fix:

cilum agent log should log correct message + "Falling back to iptables-based masquerading."

Signed-off-by: Vincent Li <v.li@f5.com>

vincentmli committed:
Refactor the code so no need to sync if and switch case check

the code is doing if and switch case check, no need to check
it twice.

Signed-off-by: Vincent Li <v.li@f5.com>

jibi reviewed (approved):
Not sure if it would be more idiomatic to make `msg` an `error` and then log it with `log.WithError(err).Warn("Falling back to iptables-based masquerading.")`?

jibi reviewed (approved):
Double approval :heavy_check_mark: :heavy_check_mark: 
Thanks for going through the refactoring and nits :pray:
